keps/sig-auth/3221-multiple-authorization-webhooks/README.md

```diff
@@ -0,0 +1,623 @@
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Today, `kube-apiserver` allows one to configure the authorization chain using
+a set of command line flags of the format `--authorization-*`. And cluster
+administrators can only enable one webhook as a part of the authorization
+chain using the `--authorization-modes` flag. This limits admins from creating
+authorization chains with multiple webhooks that validate requests in a certain
+order. This proposal makes the case for a more structured config to define the
+authorization chain for `kube-apiserver` and allowing multiple webhooks with
+well-defined parameters and enabling fine grained control like explicit Deny
+authorizer and more definitive `SubjectAccessReviews`.
+
+## Motivation
+
+Cluster administrators should be able to specify more than one authorization
+webhook in the API Server handler chain. They also need to be able to
+declaratively configure the authorizer chain using a configuration file. It
+should also be easy to say when to Deny requests, for example, when a webhook
+is unreachable.
+
+### Goals
+
+- Define a configuration file format for configuring Kubernetes API Server
+Authorization chain.
+- Allow ordered definition of authorization modes.
+- Allow definition of multiple webhooks in the authorization chain.
+- Enable user to define the policy when a webhook can't be reached due to
+network errors etc.
+- Enable more definitive results for a `SubjectAccessReview` by allowing
+config to distinguish webhooks which might deny access.
+
+### Non-Goals
+
+- Allow resource/user based pre-filtering of webhooks to prevent unnecessary
+invocations.
+- Replace path to kubeconfig file for webhooks by allowing auth configuration
+within the new file structure
+- Hot reload of the configuration.
+- Allow ordering of super user authorizer (always allow user group;
+`system:masters`) so that it can be moved from the beginning of the auth chain.
+
+> Note: The above may be revisited in future alpha iterations of this KEP or as a separate KEP (for e.g., the superuser authorizer)
+
+## Proposal
+
+Add a configuration format having specific precedence order and defined failure modes for configuring authorizer chain.
+
+```yaml
+apiVersion: apiserver.config.k8s.io/v1alpha1
+kind: AuthorizationConfiguration
+authorizers:
+- type: Webhook
+  webhook:
+    kubeConfigFile: <path/to/kubeconfigfile>
+    authorizedTTL: 5m
+    unauthorizedTTL: 30s
+    timeout: 3s
```

Comment:
Since the duration value has the unit along with the value, it is fine to not have the unit in the field name. Also, since the API is not served over HTTP, there would be no ambiguity between time duration parsers, which is why we ask in the API convention that field names should have units.

Reply:
I'm a bit worried that someone will put eg `unauthorizedTTL: 500m` intending to specify half a second, and get the wrong outcome. How do we avoid that?

Reply:
This is going into the territory of how do we ensure admins always put correct values. If they did intend to specify half a second, they should use 500ms.

Reply:
> _If there are two or more ways to do something, and one of those ways can result in a catastrophe, then someone will do it._

If we can confirm that misconfiguration due to misunderstanding can't have serious adverse consequences, I'm happy to consider this resolved. Otherwise, my concern would remain.

Reply:
We can't prevent this misconfiguration even if we explicitly add units to the field. The only way would be to put limits to the parsed duration. For the `timeout` field, we do have a max set to 30s.